subsys/rtio/rtio_executor.c

```diff
@@ -122,71 +125,108 @@ void rtio_executor_submit(struct rtio *r)
 /**
  * @brief Handle common logic when :c:macro:`RTIO_SQE_MULTISHOT` is set
  *
- * @param[in] r RTIO context
- * @param[in] curr Current IODev SQE that's being marked for finished.
- * @param[in] is_canceled Whether or not the SQE is canceled
+ * @param[in] iodev_sqe IODEV SQE that's being marked as finished.
+ * @param[in] result The result of the latest request iteration
+ * @param[in] is_ok Whether or not the SQE's result was successful
  */
-static inline void rtio_executor_handle_multishot(struct rtio *r, struct rtio_iodev_sqe *curr,
-						  bool is_canceled)
+static inline void rtio_executor_handle_multishot(struct rtio_iodev_sqe *iodev_sqe,
+						  int result, bool is_ok)
 {
-	/* Reset the mempool if needed */
-	if (curr->sqe.op == RTIO_OP_RX && FIELD_GET(RTIO_SQE_MEMPOOL_BUFFER, curr->sqe.flags)) {
-		if (is_canceled) {
-			/* Free the memory first since no CQE will be generated */
-			LOG_DBG("Releasing memory @%p size=%u", (void *)curr->sqe.rx.buf,
-				curr->sqe.rx.buf_len);
-			rtio_release_buffer(r, curr->sqe.rx.buf, curr->sqe.rx.buf_len);
-		}
+	if (!is_ok) {
+		/** Error handling of multi-shot submissions require
+		 * stopping re-submitting if something goes wrong.
+		 * Let the application decide what's best for handling
+		 * the corresponding error: whether re-submitting,
+		 * rebooting or anything else.
+		 */
+		iodev_sqe->sqe.flags |= RTIO_SQE_CANCELED;
```

Comment:
Why do we mark an errored sqe as cancelled?

Reply:
I'm a bit lost on this. This branch doesn't seem to be needed unless I've missed something.

All this flag is going to do is ensure the if (is_cancelled) branch below takes place, which reads wrong to me. The branch below should be reading...

If the submission has been cancelled *or* has failed, we free the resources. We should be doing that *before* notifying the completion.

Imagine I do

```C
RTIO_DEFINE(r, 1, 1)
```

submit a multishot, get the completion (errored), and now try and submit here, I won't have the submission available to me.

Is there some other place I'm not seeing that this flag is looked at for multishot submissions?

Reply:
Resolved to just release resources on error, without cancelling (as suggested) If we need to come back to this we can revisit.

Reply:
Ah you are thinking of the handles case. With user mode the thread won’t see updates to the sqe as it’s a copy. Only the user data connects the sqe and cqe together, really cancelled is a bit odd and changed a lot of things to make the sqe mutable instead of immutable as they were originally